### PR TITLE
Changed parser in `parse_azfp.py` to maintain consistency over other parsers.

### LINKED ISSUE #659: Improve parsing config XML in AZFP files
Currently only a subset of variables are parsed from the AZFP XML files. These are the ones that were parsed in the reference Matlab code from the manufacturer ASL Env Sci.

We recently found that in order to encode a unique `channel` identifier, it would be the best to use the field `Frequency Number` in the AZFP XML file. However, the current simplistic XML parsing function does not support that. Let's take this chance to improve the robustness of this XML parsing, and also to include the variable units which were not included in the current parser.

For needing a unique `channel` identifier, see #566, which is addressed by #657.

This is similar to #468.

## Tasks
- [x] improve robustness of AZFP XML parser, using something like [lxml](https://lxml.de/)
- [x] improve AZFP XML parser to include all config variables
- [ ] include variable units in the XML file to parser and store accordingly in the set_groups methods

### PR DESCRIPTION
**Pull request by praneethratna**
Fixes #659. Now `parse_azfp.py` uses `xml.etree.ElementTree` instead of `xm.dom.minidom` for parsing.

Update: More parameters are being parsed now, and some of those were added to the EchoData object via set_groups.

CC @leewujung @emiliom

### CODE REVIEW

File: echopype/convert/parse_azfp.py
Comment (lsetiawan) [inline echopype/convert/parse_azfp.py:134]: Hi @praneethratna! Thanks for this much cleaner implementation. I do have a question, does `ET.parse` allow for remote files? so if my xml is located at `http://mysite.com/myxml.xml` would this read and load that xml into an Element Tree object?

Comment (praneethratna) [reply echopype/convert/parse_azfp.py:134]: @lsetiawan I don't think so, i missed this point - we have to use `fsspec` and pass `xmlmap.fs.open(xmlmap.root)` as input to `ET.parse`.

Comment (emiliom) [reply echopype/convert/parse_azfp.py:134]: Yeah, it looks like you should retain the `xmlmap` statement (line 34), then, as you said, pass `xmlmap.fs.open(xmlmap.root)` as input to `ET.parse`. You can probably use a single, longer statement:
```python
root = ET.parse(xmlmap.fs.open(xmlmap.root)).getroot()
```

File: echopype/convert/parse_azfp.py
Comment (emiliom) [inline echopype/convert/parse_azfp.py:132]: This new statement doesn't make use of the `element` argument, which is required by the call in line 146.

Comment (praneethratna) [reply echopype/convert/parse_azfp.py]: @emiliom I removed the usage of this function for parsing `AZFP` XML files.

**Comment by praneethratna** [comment]
Closing and re opening to re-run the tests.

File: echopype/convert/parse_azfp.py
Comment (leewujung) [inline echopype/convert/parse_azfp.py:99]: What do you think about changing `child.tag + key` to `child.tag + key.capitalize()` so that we have the same camelcase form that is then put through the same converting camelcase to lowercase+"_" procedure, so that at the output end in `self.parameters` we have identical strings? For example, right now the first key would be `InstrumentTypestring`, and if we do the above, we will get `instrument_type_string` -- this latter one would go very well with `instrument_type` for cross referencing what is what.

File: echopype/convert/parse_azfp.py
Comment (leewujung) [inline echopype/convert/parse_azfp.py:115]: you could probably wrap these in a subfunction so that you can use here and also in the above for the `child.tag + key.capitalized` fields ([this comment](https://github.com/OSOceanAcoustics/echopype/pull/1135/files?diff=unified&w=0#r1328983167)).

File: echopype/tests/convert/test_convert_azfp.py
Comment (leewujung) [inline echopype/tests/convert/test_convert_azfp.py:179]: The test looks great! Could you add more tests for the values, especially those that are across all 4 frequencies? Just so that we have a piece of mind that things are 100% correct. Thanks!

File: echopype/convert/parse_azfp.py
Comment (praneethratna) [reply echopype/convert/parse_azfp.py:99]: Sounds great!, but one query - `sensors_flag` has `PressureSensorInstalled` as one attribute, should that be converted to `sensors_flag_pressure_sensor_installed` or `sensors_flag_pressuresensorinstalled`? What do you think about this? I think the latter one should be correct.

File: echopype/convert/set_groups_azfp.py
Comment (praneethratna) [reply echopype/convert/set_groups_azfp.py:673]: Oh sorry, I accidentally removed those changes when i encountered merge conflicts after raising the PR. Will add them back in the next commit!

**Review by leewujung — COMMENTED** [review]
@praneethratna : Thanks for the PR, it's awesome to see this very clean handling of the parameters/parameter names!! Once the few small things I commented on are included, we can merge this.

I opened up a new issue for additional work in `set_groups_azfp` to store parameters that are  potentially varying across the "phases":  https://github.com/OSOceanAcoustics/echopype/issues/1171
Let me know if you'd like to contribute to that!

File: echopype/convert/parse_azfp.py
Comment (leewujung) [reply echopype/convert/parse_azfp.py:99]: hmmm, this is a tricky question. I think `sensors_flag_pressure_sensor_installed` is easier to read. What if we go with that one?

Comment (praneethratna) [reply echopype/convert/parse_azfp.py:99]: Sure, we can go with `sensors_flag_pressure_sensor_installed`. This naming choice does make it more clear and readable.

**Comment by praneethratna** [comment]
> I opened up a new issue for additional work in `set_groups_azfp` to store parameters that are potentially varying across the "phases": #1171 Let me know if you'd like to contribute to that!

Sure, would love to contribute to #1171 once this is merged.

File: echopype/tests/convert/test_convert_azfp.py
Comment (praneethratna) [reply echopype/tests/convert/test_convert_azfp.py:179]: Added new tests for other parameters!

File: echopype/convert/parse_azfp.py
Comment (emiliom) [inline echopype/convert/parse_azfp.py]: Actually, you're converting *from* CamelCase *to* snake_case; so, the function name should be changed.

More generally, conversion of strings between CamelCase and snake_case is a really common need, in Python and all languages. I was curious about existing implementations in Python (I thought it might even be in a standard library, but it wasn't). But -- unsurprisingly :stuck_out_tongue_winking_eye: -- it's already done elsewhere in echopype, for the EK parser:
https://github.com/OSOceanAcoustics/echopype/blob/3ca028f04a4fd3643bb0ad350dc09a699883ee1e/echopype/convert/utils/ek_raw_parsers.py#L709

There, it's wrapped up within a class.

Later, in a follow-up PR, @leewujung I think we should create a generic, stand-alone utility function to do this, then reuse it both here and with EK's. For this PR, I think it'd extend its scope too much. In the meantime, I think the EK implementation is more comprehensive. I tested that function and the one implemented here. For cases like "firstLast" (first character is not uppercase) and "FirstSecond_last" (mixed form), your function drops the parts that are not strictly CamelCase and returns "last" and "first_second", respectively. The `from_CamelCase` EK function returns "first_last" and "first_second_last", as expected. I found an [even more compact implementation](https://rodic.fr/blog/camelcase-and-snake_case-strings-conversion-with-python/) that also behaves like `from_CamelCase`. Your implementation also throws away numbers, or worse.

My recommendation for this PR: copy the `from_CamelCase` function `convert/utils/ek_raw_parsers.py` exactly as is, here. Note that there's no need for it to have the `self` argument, since it's not used.

Having identical functions will make it easier for the next PR that will move the functions to a utility function.

**Comment by leewujung** [comment]
Thanks a lot @praneethratna! I did a direct push for some cosmetic changes which you can see above, and will go ahead to merge this once the tests run through. Excited to have this upgrade in!

File: echopype/convert/parse_azfp.py
Comment (emiliom) [reply echopype/convert/parse_azfp.py]: Looks like I submitted my comment just as you were renaming and moving the function!

Comment (leewujung) [reply echopype/convert/parse_azfp.py]: @emiliom : I was about to merge this. Could you do what you said above, but instead of copying that function, move that function from `ek_raw_parsers.py` to `echopype/utils/misc.py`? We'll add other things to that module later, for example from #1124. After you add that I'll merge this PR. Thanks.

File: echopype/convert/parse_azfp.py
Comment (emiliom) [inline echopype/convert/parse_azfp.py:1]: Now that `PulseLen` is auto-converted to `pulse_len` and `LockOutIndex` to `lock_out_index`, it seems like other, existing usages of the old names `pulse_length` and `lockout_index` would need to be updated, too? For example, the `field_w_freq` assignments in `_split_header`.

File: echopype/convert/parse_azfp.py
Comment (emiliom) [reply echopype/convert/parse_azfp.py]: Will do. I'm working on it.

File: echopype/convert/parse_azfp.py
Comment (praneethratna) [reply echopype/convert/parse_azfp.py:1]: Missed that, will update them.

Comment (praneethratna) [reply echopype/convert/parse_azfp.py:1]: @emiliom I pushed changes for this.

Comment (emiliom) [reply echopype/convert/parse_azfp.py:1]: Thanks! I'm about to push the changes to the definition and use of the CamelCase > snake_case function.

Comment (praneethratna) [reply echopype/convert/parse_azfp.py:1]: @emiliom Cool, thanks!

**Comment by emiliom** [comment]
@praneethratna see the new test errors. They look like they'll be easy to fix.

**Comment by praneethratna** [comment]
> @praneethratna see the new test errors. They look like they'll be easy to fix.

I actually thought about this error but forgot while merging😅. I've fixed the error and pushed the changes!

File: echopype/echodata/sensor_ep_version_mapping/v05x_to_v06x.py
Comment (emiliom) [inline echopype/echodata/sensor_ep_version_mapping/v05x_to_v06x.py:706]: Actually, leave this module as is, w/o changes. That's because it's intended only to address the conversion of files from a very old version of echopype (0.5) to another old version (0.6)
```suggestion
            "VTX",
```

**Comment by praneethratna** [comment]
@emiliom I'm sorry for all these messed up commits, kept forgetting minor changes in code. I think this can be merged as the failing test case is not related this PR.

**Comment by emiliom** [comment]
> @emiliom I'm sorry for all these messed up commits, kept forgetting minor changes in code. I think this can be merged now as the failing test is not related this PR.

Can you elaborate? I realize it's only happening with Python 3.11, and I can see that it looks like the same TypeError in `echopype/tests/convert/test_parsed_to_zarr.py`: "TypeError: 'numpy.float64' object cannot be interpreted as an integer". But if it wasn't happening before this PR, it seems like it's linked to this PR. Do you remember if it's been happening with 3.11 since you first created the PR?

Are you set up to [run tests locally](https://echopype.readthedocs.io/en/stable/contributing.html)? If not, I can try running `test_parsed_to_zarr.py` locally in a 3.11 conda environment.

**Comment by praneethratna** [comment]
> > @emiliom I'm sorry for all these messed up commits, kept forgetting minor changes in code. I think this can be merged now as the failing test is not related this PR.
> 
> Can you elaborate? I realize it's only happening with Python 3.11, and I can see that it looks like the same TypeError in `echopype/tests/convert/test_parsed_to_zarr.py`: "TypeError: 'numpy.float64' object cannot be interpreted as an integer". But if it wasn't happening before this PR, it seems like it's linked to this PR. Do you remember if it's been happening with 3.11 since you first created the PR?

Yes, it's been happening since i first created this PR.

> Are you set up to [run tests locally](https://echopype.readthedocs.io/en/stable/contributing.html)? If not, I can try running `test_parsed_to_zarr.py` locally in a 3.11 conda environment.

There seems to be an issue with setting up conda in my laptop, so I don't use any conda environment while running the tests.

**Comment by emiliom** [comment]
> Yes, it's been happening since i first created this PR.

Hmm, that's odd. I'll take a look locally. In the meantime, how about if you start planning for #1171?
Thanks.

**Comment by emiliom** [comment]
All the tests in `echopype/tests/convert/test_parsed_to_zarr.py` run fine locally with Python 3.11, with a freshly created conda environment, on both `dev` and an older version of this PR. So, I don't know what's going on, but it looks like it's limited to the CI.

I'll take a final look at the PR to confirm that everything has been addressed.

**Review by emiliom — APPROVED** [review]
It looks good! I can merge it.

But first I'll modify the PR description, b/c it's become broader than just changing how the XML is parsed

**Pull request merged**
